### DISCUSSION TITLE
Can we see actual live settings for most hardware (CPU and RAM)?

### DISCUSSION THREAD

**Discussion by OneOfMany07** [memtest86plus/memtest86plus#165] [Enhancement]
The bandwidth for caches and RAM appears tested at runtime, but the rest of the information is from defaults not the current running settings.  Is it possible to show current settings instead?

Specifically the expected max turbo speed for the CPU (based on threads setting), and live RAM values (bus speed + latencies).

Currently I see the rated non-turbo CPU speed (4.0GHz vs. my actual 4 core turbo of 4.5GHz or the default 4.2GHz even) and the XMP setting from the SPD (800MHz/1600MHz CAS 9 9 9 49 vs. live 2000Mhz CAS 10 11 11 ...).

For me...my BIOS doesn't show me the RAM latency values it'll use when I set the memory timings to "auto".  It lists the XMP defaults, even when I've turned off the XMP profile and manually increased the frequency.  CPU-z shows actual runtime values, but it'd be nice to have verification of what I'm testing from inside memtest86+ itself (maybe to include in any saved test results too?)

I'm sure this would be lower priority than a lot of other work, but I wanted to bring up the idea in case it's easy or it impacts other work.

**Comment by debrouxl** [comment]
Thanks for reaching out.

* since the methods (MSRs, MMIO ranges, etc.) to access IMCs are documented, implementing and testing IMC support foundation, then support for any given _single_ platform, so as to see the real memory settings, is doable, without spending time performing reverse-engineering work. The difficulty and amount of work come from adding support for all relevant IMC models from the past couple decades or so, and of course testing support for every IMC on at least one piece of real hardware which contains that model;
* CPU turbo mode (or EDAC, for receiving and displaying information about corrected errors, #92 ) support are affected by the same breadth issue, though the methods to achieve them are documented as well.

So it is indeed possible, but it represents a lot of work, so as you guessed, it's certainly lower priority than a number of other tasks :)
Thinking out loud: once the foundations are there, implementing IMC + EDAC + CPU turbo mode support for any given platform together at the same time would help reducing testing time, but maybe not developer time, at least at first ?

**Comment by x86fr** [comment]
This is already planned in the code but in the TODO list for now.

https://github.com/memtest86plus/memtest86plus/blob/0f8981412c49c630c312a086687878c89275d0df/app/display.c#L266-L284

Accessing IMCs controller is complex and I disagree with @debrouxl : more and more registers are undocumented on the public datasheets and requires access to NDA datasheets. Anyway, it is planned for Memtest86+ 6.10, starting with the newer platforms like AM5 and LGA1700. 😀

**Comment by debrouxl** [comment]
Right... well, NDA datasheets are a form of documentation, but unsuitable for _published_ GPL'ed software such as memtest86+, until the NDA somehow becomes void, yeah :)

**Comment by yoshinoyatoko** [comment]
SMBIOS Type-17 has one field  to tell current dram running freq

**Reply by x86fr** [reply]
SMBIOS is not a reliable source. You have to check IMC registers to be sure.
